bothan-core/src/manager/crypto_asset_info/price/tasks.rs

```diff
@@ -0,0 +1,156 @@
+                warn!("error while processing signal id {}: {}", id, e);
+                cache.set_unavailable(id);
+            }
+            Err(Error::FailedToProcessPostSignal(e)) => {
+                warn!("error while post processing signal id {}: {}", id, e);
+                cache.set_unavailable(id);
+            }
+        }
+    }
+
+    ids.iter()
+        .map(|id| cache.get(id).cloned().unwrap()) // This should never fail as all values of ids should be inserted into the cache
+        .collect()
+}
+
+async fn compute_signal_result<'a>(
+    id: &str,
+    workers: &WorkerMap<'a>,
+    registry: &Registry<Valid>,
+    stale_cutoff: i64,
+    cache: &PriceCache<String>,
+) -> Result<Decimal, Error> {
+    match registry.get(id) {
+        Some(signal) => {
+            // Check if all prerequisites are available, if not add the missing ones to the queue
+            // and continue to the next signal
+            let prereq_values = get_prerequisites(signal, cache)?;
+
+            let mut source_results = Vec::with_capacity(signal.source_queries.len());
+            for source_query in &signal.source_queries {
+                if let Some(w) = workers.get(&source_query.source_id) {
+                    let query_id = &source_query.query_id;
+                    let source_id = &source_query.source_id;
+                    match w.get_asset(&source_query.query_id).await {
+                        Ok(AssetState::Available(a)) => {
+                            if a.timestamp.ge(&stale_cutoff) {
+                                let price = compute_source_routes(
+                                    a.price,
+                                    &source_query.routes,
+                                    &prereq_values,
+                                );
+                                source_results.push(price);
+                            } else {
+                                warn!("asset {query_id} from {source_id} is stale",);
+                            }
+                        }
+                        Ok(_) => {
+                            warn!("asset state for {query_id} from {source_id} is unavailable")
+                        }
+                        Err(_) => warn!("error while querying source {source_id} for {query_id}"),
+                    }
+                }
+            }
+
+            let processed_signal = signal.processor.process(source_results)?;
+            let post_processed_signal = signal
+                .post_processors
+                .iter()
+                .try_fold(processed_signal, |acc, post| post.process(acc))?;
+
+            Ok(post_processed_signal)
+        }
+        None => Err(Error::InvalidSignal),
+    }
+}
+
+fn get_prerequisites(signal: &Signal, cache: &PriceCache<String>) -> Result<Vec<Decimal>, Error> {
+    let pids = signal
+        .source_queries
+        .iter()
+        .flat_map(|s| s.routes.iter().map(|r| r.signal_id.clone()))
```

Comment:
why do we flat map all the routes? i think this is confusing and error prone. I think this has bugs too as `compute_source_routes` does not have relationship between different `source_queries`

Reply:
resolved